Update exercices to use Effect instead of fp-ts

kevintc opened the pull request:
/!\ This is a simple _translation_ of the exercices, it could benefit from a broader view on how to implement other concepts from Effect that are not part of the current exercices.

How to review:
- Consider each exercise
- See if the translation from `fp-ts` to `Effect` is correct
- Suggest improvements and fixes if needed

kevintc committed:
Update exo0 to Effect

kevintc committed:
Update exo2 to Effect

kevintc committed:
Update exo3 to Effect

kevintc committed:
Update exo4 to Effect

kevintc committed:
Update exo5 to Effect

kevintc committed:
Update exo6 to Effect

kevintc committed:
Update exo7 to Effect

kevintc committed:
Add Effect version of utils

kevintc committed:
Update exo1 to Effect

kevintc committed:
Add Cursor rules

kevintc committed:
Add missing testUtils update

kevintc committed:
Update project dependencies and Name

Adds Effect dependencies

LaureRC commented on src/exo4 - Dependency injection with Reader/exo4.effect.exercise.ts:
```diff
@@ -0,0 +1,97 @@
+// proper exclamation style.
+
+export const exclamation = (sentence: string): Effect.Effect<string, never, CountryService> => {
+  return Effect.gen(function* (_) {
```
FYI, this is deprecated: https://effect.website/docs/getting-started/using-generators/#adapter
You should just write:
```ts
Effect.gen(function* () {
  // ...
})
```

LaureRC commented on package.json:
```diff
@@ -1,7 +1,7 @@
 {
-  "name": "fp-ts-training",
+  "name": "Effect training",
```
```suggestion
  "name": "effect-training",
```
See doc: https://docs.npmjs.com/cli/v10/configuring-npm/package-json#name -> the "name" key must not contain spaces

LaureRC commented on src/exo0 - Composing with pipe and flow/exo0.effect.exercise.ts:
```diff
@@ -0,0 +1,81 @@
+// `pipe` or `flow`), write the function `isOdd` that checks if a number is
+// odd.
+
+export const isOddP = (value: number) => pipe(value, isEven, not)
```
Seems like you solve the exercise in the file where you're supposed to keep `unimplemented` methods! :)

LaureRC commented on src/exo1 - Basic types/exo1.effect.solution.ts:
```diff
@@ -0,0 +1,117 @@
+//
+// BONUS: Try now to re-write `safeDivide` without any `if`
+//
+// HINT: Have a look at `fromPredicate` constructor
```
```suggestion
// HINT: Have a look at `liftPredicate` constructor
```
This is the equivalent of `fromPredicate` in Effect

LaureRC commented on src/exo1 - Basic types/exo1.effect.solution.ts:
```diff
@@ -0,0 +1,117 @@
+// HINT: Have a look at `fromPredicate` constructor
+
+export const safeDivideBonus = (a: number, b: number): Option.Option<number> => {
+  // First create an Option from b, ensuring it's not zero
+  const nonZeroB = pipe(
+    Option.some(b),
+    Option.filter((n: number) => n !== 0)
+  );
+  
+  // Then map it to perform the division
+  return pipe(
+    nonZeroB,
+    Option.map((b) => a / b)
+  );
```
And then: 

```suggestion
  pipe(
    b,
    Option.liftPredicate(n => n!= 0),
    Option.map(b => a / b)
  )
```

FredericEspiau commented:
Sure you don't want to either:

- keep both versions (`fp-ts` and `Effect`)
- create a whole new repository for `Effect` and keep this one as an archive ?

Because I've often heard people talk about this repo so to me it sounds look like a bit of a landmark for `Inato`

LaureRC commented on src/exo1 - Basic types/exo1.effect.solution.ts:
```diff
@@ -0,0 +1,117 @@
+///////////////////////////////////////////////////////////////////////////////
+
+// Write the safe version of `divide` with signature:
+// safeDivideWithError : (a: number, b: number) => Either<DivideByZeroError, number>
```
In Effect, the type are reversed (the value is shown first, then error, then dependencies), so:

```suggestion
// safeDivideWithError : (a: number, b: number) => Either<number, DivideByZeroError>
```

LaureRC commented on src/exo1 - Basic types/exo1.effect.solution.ts:
```diff
@@ -0,0 +1,117 @@
+// BONUS POINT: Implement `safeDivideWithError` in terms of `safeDivide`.
+//
+// HINT : Either has two basic constructors:
+// - `E.left(leftValue)`
+// - `E.right(rightValue)`
+// as well as "smarter" constructors like:
+// - `E.fromOption(() => leftValue)(option)`
```
```suggestion
// - `Either.left(leftValue)`
// - `Either.right(rightValue)`
// as well as "smarter" constructors like:
// - `Either.fromOption(() => leftValue)(option)`
```

LaureRC replied:
Also I'm not sure when to mention them, but dual API in effect are frequent, so here you could use:
```ts
Either.fromOption(option, () => leftValue)
```
Which is clearer IMO. But let's do that step by step 😄

LaureRC commented on src/exo1 - Basic types/exo1.effect.solution.ts:
```diff
@@ -0,0 +1,117 @@
+};
+
+// Write the safe version of `asyncDivide` with signature:
+// asyncSafeDivideWithError : (a: number, b: number) => Effect.Effect<never, DivideByZeroError, number>
```
```suggestion
// asyncSafeDivideWithError : (a: number, b: number) => Effect.Effect<number, DivideByZeroError, never>
```
And FYI you don't need to specify the last `never` as it's the default

LaureRC commented on src/exo1 - Basic types/exo1.effect.solution.ts:
```diff
@@ -0,0 +1,117 @@
+// Note: In Effect, Effect.Effect<A, E, R> where:
+// - A is the success type
+// - E is the error type
+// - R is the environment type
```
Here: https://effect.website/docs/getting-started/the-effect-type/ they mention "required dependencies" or "requirements" ? Might be clearer than "environment type"

kevintc commented:
> Sure you don't want to either:
> 
> * keep both versions (`fp-ts` and `Effect`)
> * create a whole new repository for `Effect` and keep this one as an archive ?
> 
> Because I've often heard people talk about this repo so to me it sounds look like a bit of a landmark for `Inato`

@FredericEspiau, we will probably split the 2 repositories, it's just convenient at first to have both versions while we work on porting the exercices from `fp-ts` to `effect`.

LaureRC commented on src/exo1 - Basic types/exo1.effect.test.ts:
```diff
@@ -0,0 +1,93 @@
+
+    it('should eventually return an Effect with error DivisionByZero if the denominator is zero', async () => {
+      // Using try/catch to handle the error case
+      try {
+        await Effect.runPromise(asyncSafeDivideWithError(25, 0));
+        // Should not reach this point
+        expect(true).toBe(false);
+      } catch (error: any) {
+        // Error is wrapped in a FiberFailure, check if it contains our expected error message
+        expect(error.toString()).toContain(DivisionByZero);
+      }
```
Let's pair on this, I think we could use `Effect.either` to check what we want without this try/catch block which is very non-standard

LaureRC replied:
FYI here is the related standard on use cases tests that should be useful: https://www.notion.so/inato/How-to-test-a-use-case-that-has-side-effects-b6f90b2233c746738f87a34b444004ba

kevintc committed:
Merge branch 'create-cursor-rules' into convert-to-Effect

kevintc committed:
Update exercices files

LaureRC commented on src/exo0 - Composing with pipe and flow/exo0.effect.exercise.ts:
```diff
@@ -0,0 +1,73 @@
+// - `flow(f) === f`
+// - `pipe(x, f) === f(x)`
+
+import { flow } from "fp-ts/lib/function"
```
remove me

LaureRC commented on src/exo1 - Basic types/exo1.effect.exercise.ts:
```diff
@@ -0,0 +1,90 @@
+// safeDivide : (a: number, b: number) => Option<number>
+//
+// HINT: Option has two basic constructors:
+// - `O.some(value)`
```
Let's use `Option`

LaureRC commented on src/exo1 - Basic types/exo1.effect.exercise.ts:
```diff
@@ -0,0 +1,90 @@
+///////////////////////////////////////////////////////////////////////////////
+
+// Write the safe version of `divide` with signature:
+// safeDivideWithError : (a: number, b: number) => Either<DivideByZeroError, number>
```
Either<number, DivideByZeroError>

LaureRC commented on src/exo1 - Basic types/exo1.effect.exercise.ts:
```diff
@@ -0,0 +1,90 @@
+export const safeDivideWithError: (
+  a: number,
+  b: number,
+) => Either.Either<DivisionByZeroError, number> = unimplemented;
```
Either.Either<number, DivisionByZeroError>

LaureRC commented on src/exo1 - Basic types/exo1.effect.exercise.ts:
```diff
@@ -0,0 +1,90 @@
+};
+
+// Write the safe version of `asyncDivide` with signature:
+// asyncSafeDivideWithError : (a: number, b: number) => Effect.Effect<never, DivideByZeroError, number>
```
Type to be reverted too

LaureRC commented on src/exo1 - Basic types/exo1.effect.exercise.ts:
```diff
@@ -0,0 +1,90 @@
+// asyncSafeDivideWithError : (a: number, b: number) => Effect.Effect<never, DivideByZeroError, number>
+//
+// HINT: Effect has a special constructor to transform a Promise<T> into
+// an Effect<never, Error, T>:
```
Same

LaureRC commented on src/exo2 - Combinators/exo2.effect.exercise.ts:
```diff
@@ -0,0 +1,236 @@
+// recommended to break those down into smaller private functions that can be
+// reused instead of doing one big implementation for each.
+//
+// HINT: `Either` has a special constructor `fromPredicate` that can accept
```
`fromPredicate` do not exist in Effect -> it's `liftPredicate`!

LaureRC commented on src/exo2 - Combinators/exo2.effect.exercise.ts:
```diff
@@ -0,0 +1,236 @@
+// HINT: Sequentially check for various possible errors is one of the most
+// common operations done with the `Either` type.
+//
+// NOTE: In Effect, Either is Either<Right, Left> (opposite to fp-ts)
```
```suggestion
// NOTE: In Effect, Either is Either<Right, Left>
```

Do we want to mention fp-ts?

LaureRC commented on src/exo2 - Combinators/exo2.effect.exercise.ts:
```diff
@@ -0,0 +1,236 @@
+// BONUS POINTS: If you properly defined small private helpers in the previous
+// section, they should be easily reused for those use-cases.
+
+export const smashOption = (character: Character): Option.Option<Damage> => {
+  if (isWarrior(character)) {
+    return Option.some(character.smash());
+  }
+  return Option.none();
+};
```
You should "unimplement" those, as this is the exercise file!

LaureRC commented on src/exo2 - Combinators/exo2.effect.solution.ts:
```diff
@@ -0,0 +1,198 @@
+
+// Helper functions for checking character types
+// Note: In Effect, Either is Either<Right, Left> (opposite to fp-ts)
+export const checkAttackerAndSmash = (attacker: Option.Option<Character>): Either.Either<Damage, NoAttackerFailure | InvalidAttackerFailure> => {
```
This exercise solution is not matching at all the one from fp-ts -> I recommend doing it manually here and reproduce more what was done on the fp-ts side and simply adapt the import (and little things like fromPredicate -> liftPredicate)

LaureRC committed:
♻️ Simpler version

LaureRC commented on src/exo5 - Nested data with traverse/exo5.effect.solution.ts:
```diff
@@ -0,0 +1,255 @@
+  // Implement this function using Effect's APIs
+  pipe(
+    optionalCountryCode,
+    Option.match({
```
Effect equivalent of `option.traverse` here would be `Effect.transposeMapOption` that I added recently in Effect 😏 let's use it! :)

kevintc committed:
Refactors exercises for clarity and consistency

- Removes unused import.
- Corrects Either type parameter order to align with the Effect library's convention.
- Introduces `Effect.transposeMapOption` for simplifying Option transformations.
- Streamlines array traversal using `Option.all` instead of manual reduction.

kevintc committed:
Merge remote-tracking branch 'origin/convert-to-Effect' into convert-to-Effect
